### ISSUE TITLE
queries.ActivePlublicKeys has consistency issues

### ISSUE THREAD

**Issue by adlerhurst** [zitadel/zitadel#6949]
There are currently multiple questions around `invalid token` error.

- https://discord.com/channels/927474939156643850/1173008599283404850
- https://discord.com/channels/927474939156643850/1174568573503553626
- integration tests are flaky

Is it possible that there are timing issues on the `not before` claim or the clock drift from the client? 

One cause that came up in a discussion with @muhlemmer was that the `not before`-claim in tokens. If the unix seconds are round up it is possible that a token is invalid for up to 499 ms if the token is created at >= hh:mm:ss.5 and gets valid at hh:mm:ss+1.

**Comment by livio-a** [comment]
We should additionally check the logging / error handling as well. The latest version wraps all underlying errors into a single `Errors.Token.Invalid (AUTH-7fs1e)`

**Comment by muhlemmer** [comment]
During the whole last week the error was always occurring during integration tests of the management API. Now that I have time to look into it, it doesn't occur anymore. I've been running the particular test in a `while true` bash loop for an hour now and it did no occur :(.

Hence, I'm just going to empty my brain on the digging done so far and I will open a PR with some inserted logging shortly after:

----

> https://discord.com/channels/927474939156643850/1173008599283404850

Uses an older zitadel version (2.39). This version is from before we started wrapping the error. There the error code is `APP-Reb32`, which originates from here:

https://github.com/zitadel/zitadel/blob/60688757faed94e967d4787b612f91d848edbad6/internal/authz/repository/eventsourcing/eventstore/token_verifier.go#L94-`L97

Digging further, there is a call to `op.VerifyAccessToken`:

https://github.com/zitadel/zitadel/blob/60688757faed94e967d4787b612f91d848edbad6/internal/authz/repository/eventsourcing/eventstore/token_verifier.go#L267

This function requires a keyset, which is basically an interface implemented by `repo.Query`, where `VerifySignature` is implemented here:

https://github.com/zitadel/zitadel/blob/60688757faed94e967d4787b612f91d848edbad6/internal/authz/repository/eventsourcing/eventstore/token_verifier.go#L313-L324

`Queries.ActivePublicKeys` does not use any triggers. I suspect there is consistency issue there, when new keys are generated for an instance and they don't show up in the projection yet. We can verify if this is true after evaluating the inserted log lines.

**Comment by muhlemmer** [comment]
And BTW, with the introspection optimization https://github.com/zitadel/zitadel/pull/6909 I introduced a new KeySet implementation that gets keys by ID directly from the eventstore and then chaches them. That implementation will be always consistent.

But that is only available from v2.42 onward, so using it as fix would make it difficult to backport to older versions.

**Comment by muhlemmer** [comment]
@livio-a , @adlerhurst : Another consistency issue report on instance keys that is most probably related: https://discord.com/channels/927474939156643850/1178641311323136020/1178641311323136020

**Comment by adlerhurst** [comment]
reported that the error still exists: https://discord.com/channels/927474939156643850/1174568573503553626/1180661447383064669

**Comment by muhlemmer** [comment]
I was working on a fix, but I figure something doesn't make sense.

`Queries.ActivePrivateSigningKey` is used by OIDC to sign tokens and uses the same projection as the `Queries.ActivePublicKeys` method. Meaning, if the `keypair.AddedEvent` is reduced the private and public key become visible at the same time in the projection tables.

If there was an consistency issue, we shouldn't have a private key to sign token to start with.

I remain with 2 suspects:
1. The time travel stuff is still present in the query and might inaccurately look a bit further in the past as the time used by cockroach. Given that clocks in docker containers tend to drift on laptops that go to stand-by, that might explain why I had the issue for a while and now not anymore. (I recreated the testing DB a couple of times already since the original investigation)
2. Since we introduced read-only transactions, might it be possible cockroach itself is inconsistent. For example in a multi-node cluster a read-only transaction might not  wait for data that is still being synced between the nodes.  As we use a connection pool, queries are effectively executed by different nodes, right?

What is you opinion @adlerhurst?

**Comment by muhlemmer** [comment]
I've reran integration test on a matrix of zitadel and database versions and I'm still unable to reproduce.


![Image](https://github.com/zitadel/zitadel/assets/5411563/4c983621-ba2f-4e36-8790-054222350038)

Zitadel v2.41.8 and v2.42 include extra logging. If those affected by this bug can post the server logs when thisken happens again, we can investigate further.

For client I would like to note and advise the following:

Keys are created per instance, on demand, when there is no active private key found on token signing. Not active means that either the instance is new and there are no keys at all, or keys are expired. If a client is obtaining public keys **before** tokens were created (user login) it will receive an empty keyset. ___They should re-retrieve keys if they see a token with a key ID they have not cached!!___ We've seen this gone wrong in the past as well and my suspicion is that this might be the case for some reporters of this issue.

**Comment by muhlemmer** [comment]
@roquie 

> As Shankar mentioned, this is not something Istiod should work around against. Your JWK management side should ensure not rotate to new keys unless all JWTs signed by old keys are expired or okay to ignore them (e.g. trigger re-authentication flow is acceptable).

Access and ID tokens have a default lifetime of 12 hours:

https://github.com/zitadel/zitadel/blob/1adfca9d281ab63beb5b0e2d031fd97ddda80291/cmd/defaults.yaml#L299-L300

Private keys are rotated every 6 hours and the Public key (used for verification) is 30 hours:

https://github.com/zitadel/zitadel/blob/1adfca9d281ab63beb5b0e2d031fd97ddda80291/cmd/defaults.yaml#L507-L508

That means that new keys become available on the first Token request after 6 hours have passed. The public key is available for a full 24 hours after the new keys are generated, which is double the time of Token lifetime. The defaults comply with the above statement.

**Comment by muhlemmer** [comment]
As for the current issue, we did a fixe regarding data consistency: https://github.com/zitadel/zitadel/pull/7080. We want to observe if that solves this issue.

The fix is part of version 2.42.2.

**Comment by muhlemmer** [comment]
The above log lines are created because we added additional logging to try to debug the current issue. This was added in PR #6976. 

https://github.com/zitadel/zitadel/blob/3483ec470d95d86c91024e6c79436cd7643693b2/internal/authz/repository/eventsourcing/eventstore/token_verifier.go#L264-L274

Basically, we try first to use the token as opaque (decrypt). If that errors, we write a warning and continue to attempt the verify the token as JWT (check the signature). If that would have failed as well, you would see an additional log line `token verifier repo: verify JWT access token`. As that isn't printing, everything seems to work fine from verification point-of-view.

**Comment by jeffmccune** [comment]
@roquie @muhlemmer Thank you for reporting this error.  I'm running into the problem roquie exactly described in [this comment](https://github.com/zitadel/zitadel/issues/6949#issuecomment-1857759664).  I have Istio v1.21.0 with a RequestAuthentication rule against an ID token issued by `ghcr.io/zitadel/zitadel:v2.46.0`  The backend database is PostgreSQL.

When I sat down at 8:49 this morning and tried to access a back end service, istio blocks the request with:

```json
{"response_code_details": "jwt_authn_access_denied{Jwks_doesn't_have_key_to_match_kid_or_alg_from_Jwt}"}
```

This is a dev environment, so over night no id tokens were likely issued in over a 12 hour period.

The request continued to fail for awhile, then at 9:10 AM it started to succeed.

Is this a new issue/bug I should file?  It's unclear reading this ticket what the solution / resolution is.

I have additional clusters I could set aside to troubleshoot this more deeply if it helps you and your team.  Please advise, thank you!

P.S. ZITADEL is great stuff, thank you.
